pkg/bench/query_runner.go

```diff
@@ -0,0 +1,274 @@
+package bench
+
+import (
+	"context"
+	"flag"
+	"math/rand"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/cortexproject/cortex/pkg/util/spanlogger"
+	"github.com/go-kit/kit/log"
+	"github.com/go-kit/kit/log/level"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/api"
+	v1 "github.com/prometheus/client_golang/api/prometheus/v1"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	config_util "github.com/prometheus/common/config"
+	"github.com/thanos-io/thanos/pkg/discovery/dns"
+	"github.com/thanos-io/thanos/pkg/extprom"
+)
+
+type QueryConfig struct {
+	Enabled           bool   `yaml:"enabled"`
+	Endpoint          string `yaml:"endpoint"`
+	BasicAuthUsername string `yaml:"basic_auth_username"`
+	BasicAuthPasword  string `yaml:"basic_auth_password"`
+}
+
+func (cfg *QueryConfig) RegisterFlags(f *flag.FlagSet) {
+	f.BoolVar(&cfg.Enabled, "bench.query.enabled", false, "enable query benchmarking")
+	f.StringVar(&cfg.Endpoint, "bench.query.endpoint", "", "Remote query endpoint.")
+	f.StringVar(&cfg.BasicAuthUsername, "bench.query.basic-auth-username", "", "Set the basic auth username on remote query requests.")
+	f.StringVar(&cfg.BasicAuthPasword, "bench.query.basic-auth-password", "", "Set the basic auth password on remote query requests.")
+}
+
+type queryRunner struct {
+	id         string
+	tenantName string
+	cfg        QueryConfig
+
+	// Do DNS client side load balancing if configured
+	dnsProvider *dns.Provider
+	remoteMtx   sync.Mutex
```

Comment:
Did this mutex used to guard a particular member variable? It's unclear to below when picking a random client what it's guarding? Everything?

Reply:
A better name would be addressMtx. It guards the array of addresses. There is a background process updating that slice and we want to be sure all access is mutex protected.